mlkem/reduce.c

```diff
@@ -41,26 +41,41 @@ int16_t montgomery_reduce(int32_t a) {
   // Replace C -> C * q in the above and estimate
   // q / 2^17 < 0.0254.
 
-  uint16_t u;
-  int16_t t;
-  // Compute a*q^{-1} mod 2^16 in unsigned representatives
-  u = (uint16_t)a * QINV;
+  // QINV == -3327 converted to uint16_t == -3327 + 65536 == 62209
+  const uint32_t UQINV = 62209;
+
+  //  Compute a*q^{-1} mod 2^16 in unsigned representatives
+  const uint16_t a_inverted =
+      (uint16_t)((((uint32_t)(a & 0xFFFF)) * UQINV) & 0xFFFF);
+
   // Lift to signed canonical representative mod 2^16.
   // PORTABILITY: This relies on uint16_t -> int16_t
   // being implemented as the inverse of int16_t -> uint16_t,
-  // which is not mandated by the standard.
-  t = (int16_t)u;
-  // By construction, the LHS is divisible by 2^16
-  t = (a - (int32_t)t * KYBER_Q) >> 16;
-  return t;
+  // which is implementation-defined (C99 6.3.1.3 (3))
+  //
+  // CBMC (rightly) objects to this implementationd-defined cast,
+  // so that particular proof obligation is suppressed here.
+#pragma CPROVER check push
+#pragma CPROVER check disable "conversion"
+  const int16_t t = (int16_t)a_inverted;
```

Comment:
Can we hide this by adding a `static inline cast_uint16_int16`?

Reply:
Possibly... let me try that.

Reply:
In this instance, I think it is OK to have the reader stumble, because it actually a very important point. The routine would still be functionally correct with `t` as `uint16`, but the bounds would be _twice as bad_.

Reply:
Yes... this works well. The non-portable code is now isolated in a single function, and montgomery_reduce() is much clearer as a result. Proofs and tests all OK.